speechbrain/core.py

```diff
@@ -1050,60 +1068,89 @@ def fit_batch(self, batch):
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+        amp = AutocastConfig.from_name(self.autocast)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        # We wrap the forward pass in no_sync to avoid
+        # unnecessary synchronization of gradients.
+        # Empirically, we found that this improves performance by a lot (2x speedup).
+        with self.no_sync():
+            with torch.autocast(
+                dtype=amp.dtype, device_type=torch.device(self.device).type,
+            ):
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+
+        with self.no_sync(not should_step):
+            # The objectives are removed from Autocast to avoid
+            # potential numerical instabilities with CTC loss, etc.
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+
+            scaled_loss = self.scaler.scale(
+                loss / self.grad_accumulation_factor
+            )
+            scaled_loss.backward()
+
+        if should_step:
+            self.optimizers_step()
 
         self.on_fit_batch_end(batch, outputs, loss, should_step)
         return loss.detach().cpu()
 
+    def check_gradients(self):
+        """ Checks if the gradients are finite. If not, it will emit a warning and set them to zero."""
+        for param in self.modules.parameters():
+            if param.requires_grad and param.grad is not None:
+                if not torch.isfinite(param.grad).all():
+                    param.grad = None
+
+    def freeze_optimizers(self, optimizers):
+        """By default, this method returns the passed optimizers.
+        Override this method if you want to freeze some optimizers
+        during training. To do so, return a of active optimizers.
+        """
+        return optimizers
+
+    def on_optimizers_step_end(self):
+        """Gets called at the end of ``optimizers_step()``.
+        Override this method to do something at the end of ``optimizers_step()``.
+        """
+        pass
+
+    def optimizers_step(self):
+
+        # 1. get the valid optimizers, i.e., the ones that are not frozen during this step
+        valid_optimizers = self.freeze_optimizers(self.optimizers_dict)
+
+        # 2. unscale the gradients of the valid optimizers
+        for opt in valid_optimizers.values():
+            self.scaler.unscale_(opt)
+
+        # 3. clip gradients
+        for opt in valid_optimizers.values():
+            torch.nn.utils.clip_grad_norm_(
+                opt.param_groups[0]["params"], self.max_grad_norm
+            )
+
+        # no need to activate this flag if you are in fp16
+        # since GradScaler is automatically handling the nonfinite gradients
+        if self.skip_nan_grad:
```

Comment:
`self.skip_nan_grad` is undefined

Reply:
Yes. But as I said, I did not implemented everything. See: https://github.com/speechbrain/speechbrain/pull/2010#issuecomment-1740850106 The idea was just to provide a higher view of what is expecting to change. I wanted to first hear from the core team.